### PR TITLE
[Design]: Proposal for more generic tag helper configuration feature

### PR DESCRIPTION
**Pull request by DamianEdwards**
@davidfowl and I came up with this as an alternative to creating separate options objects for tag helpers. Means you can very simply add default configuration for any tag helper without the need for an options object by just setting the properties directly from your config delegate. Properties set from source CSHTML still win. e.g. 

``` C#
services.InitializeTagHelper<FormTagHelper>((helper, viewContext) => helper.AntiForgery = false);
services.InitializeTagHelper<CacheTagHelper>((helper, viewContext) =>
{
    // Set default caching expiration to sliding 5 minutes
    helper.ExpiresAfter = TimeSpan.FromMinutes(5);
});
```

- Doesn't use IOptions
- Any Action<ITagHelper, ViewContext> can be registered in DI to allow setting of tag helper properties before their bound from the source document and the tag helper is executed
- #1689

@NTaylorMullen @dougbu @Eilon

### CODE REVIEW

**Comment by rynowak** [comment]
The tradeoff here seems to be that this code is going to execute for every instance of the taghelper used. Do you anticipate any cases where data used here is going to come non-trivial code? (Db, Configuration). This is much more pay-for-play than our other options-ey things.

What's interesting here is access to the view context this becomes much more like a filter.

**Comment by dougbu** [comment]
will this configuration story relate in any way to the `IOptions` story?  I don't see how e.g. customers could configure `<form/>` tag helpers using a config.json file.

**Comment by DamianEdwards** [comment]
@haok

**Comment by DamianEdwards** [comment]
@rynowak really depends on the app, but I could imagine some scenarios. More likely you'd want the `HttpContext` so you could features on and off by query string or cookie. Accessing the user db might be interesting for A/B testing scenarios, but that could also be done by just setting a cookie and checking that.

@dougbu we could integrate them, but right now you could still set it from config by just assigning the property in the delegate from an `IConfiguration`. The auto-binding from config feature is a little crappy right now anyway, e.g. doesn't work on .NET Core. That said, we could try to make this work via options, e.g. Putting something like `IOptions<ITagHelperOptions<FormTagHelper>>` in the container

**Comment by HaoK** [comment]
Wasn't the main motivation behind splitting settings into an options POCO to standardize that as the new pattern to configure services?  Early on @divega and I suggested removing the new() restriction on Options and activating options from DI (which would have enabled this kind of scenario), but @lodejard felt strongly that options should just be data.  This proposal seems to move back towards configuring an instance instead of setting up a singleton Options, with all service instance just using the same Options.

**Comment by NTaylorMullen** [comment]
I don't feel like this is a replacement for the current configuration system, per-instance seems overkill to make this a full replacement. If anything it should be an addition.

**Comment by DamianEdwards** [comment]
@HaoK you're correct that is more configuring an instance of a thing. The argument here is that tag helpers are different/special. They're used many times in an app but not always owned by the app and generally have a bunch of properties that directly govern their behavior. This is a pattern to enable a very simple, generic way to set properties on tag helpers at runtime (so you can easily do default conditional behavior) without the tag helper itself explicitly needing to support it. It's more akin to the skinning feature in Web Forms than options.

@NTaylorMullen per instance enables more interesting scenarios, e.g. conditional defaults based on the current request. But like i said, we could look at basing this same thing on `IOptions` without losing the per-instance capability.

**Comment by HaoK** [comment]
Yeah I always felt configuring instances using the same mechanism as Options is useful, perhaps what's needed is exposing the options style configuration at a more basic level, so we could use it both to configure instances like you are doing, and also to configure options like we are doing today...  Basically its the `Configure<MyThingy>, (me => me.Whatever = "yolo";` syntax that seems nice in general (for both instances/options)

**Comment by NTaylorMullen** [comment]
Design looks good, only semi-design comment is you should probably cache the `configureEnumerableType` in your `DefaultTagHelperActivator` since we'll be activating on a per-tag helper on the page basis.

**Comment by NTaylorMullen** [comment]
Found an alternate way to pass through the `TTagHelper` to enable more magic and cleanness.

File: src/Microsoft.AspNet.Mvc.Razor/ConfigureTagHelperOfT.cs [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+        /// <summary>
+        /// Configures the <see cref="TTagHelper"/> using <see cref="Action"/>;
+        /// </summary>
+        /// <param name="helper">The <see cref="TTagHelper"/> to configure.</param>
+        /// <param name="context">
+        ///     The <see cref="ViewContext"/> for the <see cref="IView"/> the <see cref="TTagHelper"/> is in.
+        /// </param>
+        public void Configure(TTagHelper helper, ViewContext context)
+        {
+            if (!helper.GetType().IsAssignableFrom(typeof(TTagHelper)))
```
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.Razor/ConfigureTagHelperOfT.cs]: no need for this check now

**Comment by DamianEdwards** [comment]
Thanks all. I think we're in a good place now. I'll add tests, etc. and send this back as a real PR.

**Pull request closed without merging**